lib/stage-copy.ts

```diff
@@ -0,0 +1,120 @@
+import * as _ from 'lodash';
+import * as path from 'path';
+import { Readable } from 'stream';
+import * as tar from 'tar-stream';
+
+import { StageCopy } from './action-group';
+import { Container } from './container';
+import { InternalInconsistencyError } from './errors';
+import { streamToBuffer } from './util';
+
+export async function copyToStage(
+	source: Container,
+	dest: Container,
+	copy: StageCopy,
+) {
+	if (await source.pathIsDirectory(copy.source)) {
+		await copyDirToStage(source, dest, copy);
+	} else {
+		await copyFileToStage(source, dest, copy);
+	}
+}
+
+async function copyDirToStage(
+	source: Container,
+	dest: Container,
+	copy: StageCopy,
+) {
+	if (!(await dest.pathIsDirectory(copy.dest))) {
+		throw new InternalInconsistencyError(
+			`Attempt to copy directory into non-directory destination from stage copy: destination: ${
+				copy.dest
+			}`,
+		);
+	}
+
+	// Get everything from the directory of the source container
+	const sourceStream = await source.fetchPathFromContainer(copy.source);
+
+	// Now we need to go through, creating a new tar stream, which
+	// sets up the paths based relative to root
+
+	const extract = tar.extract();
+	const pack = tar.pack();
+
+	await new Promise((resolve, reject) => {
+		extract.on(
+			'entry',
+			async (headers: tar.Headers, stream: Readable, next: () => void) => {
+				let pathName = headers.name;
+				if (path.isAbsolute(copy.source)) {
+					pathName = `/${pathName}`;
+				}
+				const toAppend = path.relative(copy.source, pathName);
+
+				pack.entry(
+					{ ...headers, name: path.join(copy.dest, toAppend) },
+					await streamToBuffer(stream),
```

Comment:
This will use some memory for large files.
Why not
```typescript
const entry = pack.entry(
	{ ...headers, name: path.join(copy.dest, toAppend) },
	err => {
		if (err) {
			reject(err);
		}
		next();
	},
);
stream.pipe(entry);
```
as you have the file size in `headers`?
https://github.com/mafintosh/tar-stream#packing

Reply:
I've just had some seriously unreliable behaviour with tar-stream and the streaming interface. I prefer to use the buffer interface for that reason. Also it will exist in memory afterwards so I think it's ok for now.